server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1132,7 +1170,8 @@ private RepositoryData safeRepositoryData(long repositoryStateId, Map<String, Bl
             cached = null;
         } else {
             genToLoad = latestKnownRepoGen.get();
-            cached = latestKnownRepositoryData.get();
+            SoftReference<Tuple<Long, BytesReference>> softRef = latestKnownRepositoryData.get();
```

Comment:
@inpink could you please educate me why the `SoftReference` is being chosen and not `WeakReference` ?

Reply:
My understand is a WeakReference will be collected even if the JVM isn't under any significant memory pressure if the object isn't otherwise being referenced. The JVM will hang on to a SoftReference using a more complicated formula, but is guaranteed to reclaim the space before OOMing.

I admit there's some risk here with a SoftReference, specifically if OpenSearch's circuit breakers prevent the JVM from ever hitting a memory pressure scenario where it will collect soft references, then these references might never be reclaimed.

Reply:
> The JVM will hang on to a SoftReference using a more complicated formula, but is guaranteed to reclaim the space before OOMing. 

Correct, this is my understanding, I think if we go with weak, the risks could be largely eliminated, right?

Reply:
> I think if we go with weak, the risks could be largely eliminated, right?

I think we'd have the opposite risk: a performance degradation because the weak reference would almost always be collected before it could be re-used.

Reply:
Fair point, so what if we go with `SoftReference` + TTL as an alternative? Indeed,  circuit breakers may trigger any time (and this snapshot metadata reference hanging around could be the cause), but with TTL we could actually check the last accessed time and drop the `SoftReference` altogether?

Reply:
I believe @andrross has provided an excellent explanation for why SoftReference was chosen. Thank you very much! We previously discussed SoftReference in this issue, so I’m sharing it here in case it might be helpful: [link to issue discussion](https://github.com/opensearch-project/OpenSearch/issues/16298#issuecomment-2445248228). Any additional thoughts you might have would be greatly appreciated.

@reta, if I understood correctly, your suggestion of a TTL was to help reduce the unpredictability associated with SoftReference—would that be accurate? This cache is accessed each time a snapshot is added, deleted, or restored. If snapshots are taken every hour over 14 days, this would mean metadata for 336 snapshots would be cached (following AWS OpenSearch’s snapshot storage approach as described [here](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/managedomains-snapshots.html)). In this case, may I ask what you would consider an appropriate TTL duration?